File: scripts/fix-icons.js
/**
 * Fix white-corner artifacts in PWA icon PNGs.
 * Flood-fills near-white pixels from each corner with full transparency.
 */
const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

const ICONS_DIR = path.join(__dirname, '..', 'icons');
const WHITE_THRESHOLD = 240;

async function fixIcon(filePath) {
  const { data, info } = await sharp(filePath)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width, height, channels } = info; // channels === 4 (RGBA)
  const buf = Buffer.from(data);

  function idx(x, y) { return (y * width + x) * channels; }

  function isNearWhite(x, y) {
    const i = idx(x, y);
    return buf[i] > WHITE_THRESHOLD &&
           buf[i+1] > WHITE_THRESHOLD &&
           buf[i+2] > WHITE_THRESHOLD &&
           buf[i+3] > 127; // opaque enough to matter
  }

  function floodFill(startX, startY) {
    const stack = [[startX, startY]];
    const visited = new Uint8Array(width * height);
    while (stack.length) {
      const [x, y] = stack.pop();
      if (x < 0 || x >= width || y < 0 || y >= height) continue;
      const vi = y * width + x;
      if (visited[vi]) continue;
      visited[vi] = 1;
      if (!isNearWhite(x, y)) continue;
      // Make fully transparent
      const i = idx(x, y);
      buf[i] = 0; buf[i+1] = 0; buf[i+2] = 0; buf[i+3] = 0;
      stack.push([x+1,y],[x-1,y],[x,y+1],[x,y-1]);
    }
  }

  floodFill(0, 0);
  floodFill(width - 1, 0);
  floodFill(0, height - 1);
  floodFill(width - 1, height - 1);

  await sharp(buf, { raw: { width, height, channels } })
    .png()
    .toFile(filePath + '.tmp');

  fs.renameSync(filePath + '.tmp', filePath);
  console.log(`  fixed: ${path.basename(filePath)} (${width}x${height})`);
}

async function main() {
  console.log(`Processing: ${ICONS_DIR}`);
  const files = fs.readdirSync(ICONS_DIR).filter(f => f.toLowerCase().endsWith('.png'));
  for (const f of files) {
    await fixIcon(path.join(ICONS_DIR, f));
  }
  console.log('Done.');
}

main().catch(err => { console.error(err); process.exit(1); });

File: scripts/fix-icons.py
"""Remove white/near-white corner pixels from PWA icon PNGs, making them transparent."""
import os
from PIL import Image

ICONS_DIR = os.path.join(os.path.dirname(__file__), '..', 'icons')
WHITE_THRESHOLD = 240  # R,G,B all above this → treat as white edge


def flood_fill_transparent(img, start_x, start_y):
    """Flood-fill from a corner pixel, replacing near-white pixels with transparent."""
    pixels = img.load()
    w, h = img.size
    stack = [(start_x, start_y)]
    visited = set()

    while stack:
        x, y = stack.pop()
        if (x, y) in visited:
            continue
        if x < 0 or x >= w or y < 0 or y >= h:
            continue
        visited.add((x, y))
        r, g, b, a = pixels[x, y]
        # Stop if pixel is already transparent or not near-white
        if a < 128:
            continue
        if r < WHITE_THRESHOLD or g < WHITE_THRESHOLD or b < WHITE_THRESHOLD:
            continue
        pixels[x, y] = (0, 0, 0, 0)
        stack.extend([(x+1, y), (x-1, y), (x, y+1), (x, y-1)])


def fix_icon(path):
    img = Image.open(path).convert('RGBA')
    w, h = img.size
    corners = [(0, 0), (w - 1, 0), (0, h - 1), (w - 1, h - 1)]
    for cx, cy in corners:
        flood_fill_transparent(img, cx, cy)
    img.save(path, format='PNG')
    print(f'  fixed: {os.path.basename(path)} ({w}x{h})')


def main():
    icons_dir = os.path.abspath(ICONS_DIR)
    print(f'Processing icons in: {icons_dir}')
    for fname in sorted(os.listdir(icons_dir)):
        if fname.lower().endswith('.png'):
            fix_icon(os.path.join(icons_dir, fname))
    print('Done.')


if __name__ == '__main__':
    main()
